vendor/k8s.io/kubernetes/pkg/master/controller/clusterauthenticationtrust/cluster_authentication_trust_controller.go

```diff
@@ -0,0 +1,516 @@
+		if err != nil {
+			return nil, err
+		}
+		data["requestheader-group-headers"], err = jsonSerializeStringSlice(authenticationInfo.RequestHeaderGroupHeaders.Value())
+		if err != nil {
+			return nil, err
+		}
+		data["requestheader-extra-headers-prefix"], err = jsonSerializeStringSlice(authenticationInfo.RequestHeaderExtraHeaderPrefixes.Value())
+		if err != nil {
+			return nil, err
+		}
+
+		data["requestheader-client-ca-file"] = string(caBytes)
+		data["requestheader-allowed-names"], err = jsonSerializeStringSlice(authenticationInfo.RequestHeaderAllowedNames.Value())
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return data, nil
+}
+
+func getClusterAuthenticationInfoFor(data map[string]string) (ClusterAuthenticationInfo, error) {
+	ret := ClusterAuthenticationInfo{}
+
+	var err error
+	ret.RequestHeaderGroupHeaders, err = jsonDeserializeStringSlice(data["requestheader-group-headers"])
+	if err != nil {
+		return ClusterAuthenticationInfo{}, err
+	}
+	ret.RequestHeaderExtraHeaderPrefixes, err = jsonDeserializeStringSlice(data["requestheader-extra-headers-prefix"])
+	if err != nil {
+		return ClusterAuthenticationInfo{}, err
+	}
+	ret.RequestHeaderAllowedNames, err = jsonDeserializeStringSlice(data["requestheader-allowed-names"])
+	if err != nil {
+		return ClusterAuthenticationInfo{}, err
+	}
+	ret.RequestHeaderUsernameHeaders, err = jsonDeserializeStringSlice(data["requestheader-username-headers"])
+	if err != nil {
+		return ClusterAuthenticationInfo{}, err
+	}
+
+	if caBundle := data["requestheader-client-ca-file"]; len(caBundle) > 0 {
+		ret.RequestHeaderCA, err = dynamiccertificates.NewStaticCAContent("existing", []byte(caBundle))
+		if err != nil {
+			return ClusterAuthenticationInfo{}, err
+		}
+	}
+
+	if caBundle := data["client-ca-file"]; len(caBundle) > 0 {
+		ret.ClientCA, err = dynamiccertificates.NewStaticCAContent("existing", []byte(caBundle))
+		if err != nil {
+			return ClusterAuthenticationInfo{}, err
+		}
+	}
+
+	return ret, nil
+}
+
+func jsonSerializeStringSlice(in []string) (string, error) {
+	out, err := json.Marshal(in)
+	if err != nil {
+		return "", err
+	}
+	return string(out), err
+}
+
+func jsonDeserializeStringSlice(in string) (headerrequest.StringSliceProvider, error) {
+	if len(in) == 0 {
+		return nil, nil
+	}
+
+	out := []string{}
+	if err := json.Unmarshal([]byte(in), &out); err != nil {
+		return nil, err
+	}
+	return headerrequest.StaticStringSlice(out), nil
+}
+
+func combineUniqueStringSlices(lhs, rhs headerrequest.StringSliceProvider) headerrequest.StringSliceProvider {
+	ret := []string{}
+	present := sets.String{}
+
+	if lhs != nil {
+		for _, curr := range lhs.Value() {
+			if present.Has(curr) {
+				continue
+			}
+			ret = append(ret, curr)
+			present.Insert(curr)
+		}
+	}
+
+	if rhs != nil {
+		for _, curr := range rhs.Value() {
+			if present.Has(curr) {
+				continue
+			}
+			ret = append(ret, curr)
+			present.Insert(curr)
+		}
+	}
```

Comment:
Why not just
```
if lhs != nil {
    present.Insert(lhs.Value()...)
}
if rhs != nil {
    present.Insert(rhs.Value()...)
}

return headerrequest.StaticStringSlice(present.List())
```

Does the input order matter or is it just so that the behavior matches the other `combine*` functions?

Reply:
commenting on changes to upstream code.  We could change the upstream, but I don't think we want to.

Reply:
pfft, I thought <drop> means our changes on top of upstream that are supposed to be dropped on new kube version rebase. My bad.

Reply:
> pfft, I thought `<drop>` means our changes on top of upstream that are supposed to be dropped on new kube version rebase. My bad.

The features came in over the course of 8 or so PRs.  made more sense to bring it all in here at once